### PR TITLE
Try normalizing the dbcore String type

### LINKED ISSUE #3773: FetchArt crashes with TypeError for particular album
I've imported hundreds of albums just fine using `fetchart` in `auto` mode, and then this one came along and crashed the import.

The issue in short:

```
  File "/usr/local/lib/python3.7/dist-packages/beetsplug/fetchart.py", line 416, in get
    self.API_ALBUMS + album.mb_releasegroupid,
TypeError: can only concatenate str (not "int") to str
```
So it seems `album.mb_releasegroupid` is an `int`. This seems like a bug.

### Problem

Running this command in verbose (`-vv`) mode:

```sh
$ beet -vv im --set=genre=Vaporwave Vektroid\ -\ Color\ Ocean\ Road/
```

Led to this problem:

```
user configuration: /media/droppie/libraries/music/.meta/beets/config.yaml
data directory: /media/droppie/libraries/music/.meta/beets
plugin paths: /opt/whatlastgenre/plugin/beets/beetsplug
Sending event: pluginload
library database: /media/droppie/libraries/music/.meta/beets/library.db
library directory: /media/droppie/libraries/music
Sending event: library_opened
Sending event: import_begin
Sending event: import_task_created
Sending event: import_task_start
Looking up: /media/droppie/data/music/Vektroid - Color Ocean Road
Tagging Vektroid - Color Ocean Road
No album ID found.
Search terms: Vektroid - Color Ocean Road
Album might be VA: False
Searching for MusicBrainz releases with: {'release': 'color ocean road', 'artist': 'vektroid', 'tracks': '6'}
Requesting MusicBrainz release acfcb884-a136-44d3-a537-65194f07bf59
primary MB release type: album
Sending event: albuminfo_received
Candidate: Vektroid - Color Ocean Road (acfcb884-a136-44d3-a537-65194f07bf59)
Computing track assignment...
...done.
Success. Distance: 0.19
Requesting MusicBrainz release 42fa86c0-c28c-4850-ba6a-b2696808dabf
primary MB release type: album
Sending event: albuminfo_received
Candidate: Vektroid - RE•SET (42fa86c0-c28c-4850-ba6a-b2696808dabf)
Computing track assignment...
...done.
Success. Distance: 0.65
Requesting MusicBrainz release 6a832aac-a80b-4db2-a1d7-dc65d92a982d
primary MB release type: album
Sending event: albuminfo_received
Candidate: Vektroid - Telnet Complete (6a832aac-a80b-4db2-a1d7-dc65d92a982d)
Computing track assignment...
...done.
Success. Distance: 0.76
Requesting MusicBrainz release 043e9481-f864-4676-9888-924f026fa3dd
primary MB release type: album
Sending event: albuminfo_received
Candidate: Vektroid - Seed & Synthetic Earth (043e9481-f864-4676-9888-924f026fa3dd)
Computing track assignment...
...done.
Success. Distance: 0.67
Requesting MusicBrainz release 12ec78f1-169a-408c-9d9e-a699901d133c
primary MB release type: broadcast
secondary MB release type(s): dj-mix
Sending event: albuminfo_received
Candidate: Vektroid - FACT Mix 619: Vektroid (Sept '17) (12ec78f1-169a-408c-9d9e-a699901d133c)
Computing track assignment...
...done.
Success. Distance: 0.65
discogs: Searching for master release 1520897
discogs: hit rate limit, waiting for 0.9732460975646973 seconds
discogs: Searching for master release 1520897
discogs: hit rate limit, waiting for 0.9869840145111084 seconds
discogs: Searching for master release 1520897
discogs: hit rate limit, waiting for 0.9847149848937988 seconds
Sending event: albuminfo_received
Candidate: Vektroid - Color Ocean Road (4820006)
Computing track assignment...
...done.
Success. Distance: 0.22
Sending event: albuminfo_received
Candidate: Vektroid - Color Ocean Road (13385176)
Computing track assignment...
...done.
Success. Distance: 0.05
Sending event: albuminfo_received
Candidate: Vektroid - Color Ocean Road (13395427)
Computing track assignment...
...done.
Success. Distance: 0.22
Evaluating 8 candidates.

/media/droppie/data/music/Vektroid - Color Ocean Road (6 items)
Sending event: before_choose_candidate
Tagging:
    Vektroid - Color Ocean Road
URL:
    https://www.discogs.com/Vektroid-Color-Ocean-Road/release/13385176
(Similarity: 95.2%) (source) (Discogs, File, 2012, US, PrismCorp)
[A]pply, More candidates, Skip, Use as-is, as Tracks, Group albums,
Enter search, enter Id, aBort, eDit, edit Candidates? 
Sending event: import_task_choice
Sending event: import_task_apply
Replacing item 2040: /media/droppie/data/music/Vektroid - Color Ocean Road/01 Shalom.mp3
Sending event: database_change
Sending event: item_removed
Replacing item 2041: /media/droppie/data/music/Vektroid - Color Ocean Road/02 Color Ocean.mp3
Sending event: database_change
Sending event: item_removed
Replacing item 2042: /media/droppie/data/music/Vektroid - Color Ocean Road/03 Seafoam Island.mp3
Sending event: database_change
Sending event: item_removed
Replacing item 2043: /media/droppie/data/music/Vektroid - Color Ocean Road/04 Sushi Plaza.mp3
Sending event: database_change
Sending event: item_removed
Replacing item 2044: /media/droppie/data/music/Vektroid - Color Ocean Road/05 Mango _ Fuji.mp3
Sending event: database_change
Sending event: item_removed
Replacing item 2045: /media/droppie/data/music/Vektroid - Color Ocean Road/06 Om Namo Ocean Road.mp3
Sending event: database_change
Sending event: database_change
Sending event: item_removed
6 of 6 items replaced
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Reimported album: added 1602800480.7681532, flexible attributes [] from album 185 for /media/droppie/data/music/Vektroid - Color Ocean Road
Reimported item added 1602800480.7766094 from item 2040 for /media/droppie/data/music/Vektroid - Color Ocean Road/01 Shalom.mp3
Reimported item flexible attributes ['track_alt', 'data_source', 'track_alt', 'data_source'] from item 2040 for /media/droppie/data/music/Vektroid - Color Ocean Road/01 Shalom.mp3
Sending event: database_change
Reimported item added 1602800480.7842014 from item 2041 for /media/droppie/data/music/Vektroid - Color Ocean Road/02 Color Ocean.mp3
Reimported item flexible attributes ['track_alt', 'data_source', 'track_alt', 'data_source'] from item 2041 for /media/droppie/data/music/Vektroid - Color Ocean Road/02 Color Ocean.mp3
Sending event: database_change
Reimported item added 1602800480.7913542 from item 2042 for /media/droppie/data/music/Vektroid - Color Ocean Road/03 Seafoam Island.mp3
Reimported item flexible attributes ['track_alt', 'data_source', 'track_alt', 'data_source'] from item 2042 for /media/droppie/data/music/Vektroid - Color Ocean Road/03 Seafoam Island.mp3
Sending event: database_change
Reimported item added 1602800480.7984622 from item 2043 for /media/droppie/data/music/Vektroid - Color Ocean Road/04 Sushi Plaza.mp3
Reimported item flexible attributes ['track_alt', 'data_source', 'track_alt', 'data_source'] from item 2043 for /media/droppie/data/music/Vektroid - Color Ocean Road/04 Sushi Plaza.mp3
Sending event: database_change
Reimported item added 1602800480.8071747 from item 2044 for /media/droppie/data/music/Vektroid - Color Ocean Road/05 Mango _ Fuji.mp3
Reimported item flexible attributes ['track_alt', 'data_source', 'track_alt', 'data_source'] from item 2044 for /media/droppie/data/music/Vektroid - Color Ocean Road/05 Mango _ Fuji.mp3
Sending event: database_change
Reimported item added 1602800480.8171701 from item 2045 for /media/droppie/data/music/Vektroid - Color Ocean Road/06 Om Namo Ocean Road.mp3
Reimported item flexible attributes ['track_alt', 'data_source', 'track_alt', 'data_source'] from item 2045 for /media/droppie/data/music/Vektroid - Color Ocean Road/06 Om Namo Ocean Road.mp3
Sending event: database_change
Set field genre=Vaporwave for /media/droppie/data/music/Vektroid - Color Ocean Road
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
Sending event: database_change
fetchart: trying source filesystem for album Vektroid - Color Ocean Road
fetchart: trying source coverart for album Vektroid - Color Ocean Road
fetchart: downloading image: https://coverartarchive.org/release/13385176/front
fetchart: not a supported image: image/x-None
fetchart: trying source coverart for album Vektroid - Color Ocean Road
fetchart: downloading image: https://coverartarchive.org/release-group/1520897/front
fetchart: not a supported image: image/x-None
fetchart: trying source itunes for album Vektroid - Color Ocean Road
fetchart: getting URL: https://itunes.apple.com/search?term=Vektroid+Color+Ocean+Road&entity=album&media=music&limit=200
fetchart: iTunes search for 'Vektroid Color Ocean Road' got no results
fetchart: trying source amazon for album Vektroid - Color Ocean Road
fetchart: trying source albumart for album Vektroid - Color Ocean Road
fetchart: trying source wikipedia for album Vektroid - Color Ocean Road
fetchart: getting URL: https://dbpedia.org/sparql?format=application%2Fsparql-results%2Bjson&timeout=2500&query=PREFIX+rdf%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F1999%2F02%2F22-rdf-syntax-ns%23%3E%0A+++++++++++++++++PREFIX+dbpprop%3A+%3Chttp%3A%2F%2Fdbpedia.org%2Fproperty%2F%3E%0A+++++++++++++++++PREFIX+owl%3A+%3Chttp%3A%2F%2Fdbpedia.org%2Fontology%2F%3E%0A+++++++++++++++++PREFIX+rdfs%3A+%3Chttp%3A%2F%2Fwww.w3.org%2F2000%2F01%2Frdf-schema%23%3E%0A+++++++++++++++++PREFIX+foaf%3A+%3Chttp%3A%2F%2Fxmlns.com%2Ffoaf%2F0.1%2F%3E%0A%0A+++++++++++++++++SELECT+DISTINCT+%3FpageId+%3FcoverFilename+WHERE+%7B%0A+++++++++++++++++++%3Fsubject+owl%3AwikiPageID+%3FpageId+.%0A+++++++++++++++++++%3Fsubject+dbpprop%3Aname+%3Fname+.%0A+++++++++++++++++++%3Fsubject+rdfs%3Alabel+%3Flabel+.%0A+++++++++++++++++++%7B+%3Fsubject+dbpprop%3Aartist+%3Fartist+%7D%0A+++++++++++++++++++++UNION%0A+++++++++++++++++++%7B+%3Fsubject+owl%3Aartist+%3Fartist+%7D%0A+++++++++++++++++++%7B+%3Fartist+foaf%3Aname+%22Vektroid%22%40en+%7D%0A+++++++++++++++++++++UNION%0A+++++++++++++++++++%7B+%3Fartist+dbpprop%3Aname+%22Vektroid%22%40en+%7D%0A+++++++++++++++++++%3Fsubject+rdf%3Atype+%3Chttp%3A%2F%2Fdbpedia.org%2Fontology%2FAlbum%3E+.%0A+++++++++++++++++++%3Fsubject+dbpprop%3Acover+%3FcoverFilename+.%0A+++++++++++++++++++FILTER+%28+regex%28%3Fname%2C+%22Color+Ocean+Road%22%2C+%22i%22%29+%29%0A++++++++++++++++++%7D%0A+++++++++++++++++Limit+1
fetchart: wikipedia: album not found on dbpedia
fetchart: trying source fanarttv for album Vektroid - Color Ocean Road
Traceback (most recent call last):
  File "/usr/local/bin/beet", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/dist-packages/beets/ui/__init__.py", line 1266, in main
    _raw_main(args)
  File "/usr/local/lib/python3.7/dist-packages/beets/ui/__init__.py", line 1253, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/usr/local/lib/python3.7/dist-packages/beets/ui/commands.py", line 955, in import_func
    import_files(lib, paths, query)
  File "/usr/local/lib/python3.7/dist-packages/beets/ui/commands.py", line 925, in import_files
    session.run()
  File "/usr/local/lib/python3.7/dist-packages/beets/importer.py", line 329, in run
    pl.run_parallel(QUEUE_SIZE)
  File "/usr/local/lib/python3.7/dist-packages/beets/util/pipeline.py", line 445, in run_parallel
    six.reraise(exc_info[0], exc_info[1], exc_info[2])
  File "/usr/lib/python3/dist-packages/six.py", line 693, in reraise
    raise value
  File "/usr/local/lib/python3.7/dist-packages/beets/util/pipeline.py", line 312, in run
    out = self.coro.send(msg)
  File "/usr/local/lib/python3.7/dist-packages/beets/util/pipeline.py", line 194, in coro
    func(*(args + (task,)))
  File "/usr/local/lib/python3.7/dist-packages/beets/importer.py", line 1511, in plugin_stage
    func(session, task)
  File "/usr/local/lib/python3.7/dist-packages/beets/plugins.py", line 143, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.7/dist-packages/beetsplug/fetchart.py", line 854, in fetch_art
    candidate = self.art_for_album(task.album, task.paths, local)
  File "/usr/local/lib/python3.7/dist-packages/beetsplug/fetchart.py", line 920, in art_for_album
    for candidate in source.get(album, self, paths):
  File "/usr/local/lib/python3.7/dist-packages/beetsplug/fetchart.py", line 416, in get
    self.API_ALBUMS + album.mb_releasegroupid,
TypeError: can only concatenate str (not "int") to str

```

### Setup

* OS: Raspbian Buster 
* Python version: 3.7.3
* beets version: 1.4.9
* Turning off plugins made problem go away (yes/no): n/a

My configuration (output of `beet config`) is:

```yaml
library: /media/droppie/libraries/music/.meta/beets/library.db
directory: /media/droppie/libraries/music
original_date: yes
per_disc_numbering: no
va_name: "\U0001F465 Various Artists"

import:
    write: no
    copy: no
    move: no
    link: yes

    resume: ask
    incremental: yes
    quiet_fallback: skip
    none_rec_action: ask
    timid: yes
    default_action: apply
    log: /var/log/beets/import.log
    languages:
        - en
    detail: no
    group_albums: no
    autotag: yes
    duplicate_action: ask
    incremental_skip_later: no
    set_fields:
        genre: "\U0001F3A7 New"

pluginpath:
    /opt/whatlastgenre/plugin/beets/beetsplug

plugins: discogs edit fetchart bucket replaygain info follow wlg types

types:
    rating: int

edit:
    itemfields: track artist title genre albumartist
    albumfields: albumartist album genre rating

fetchart:
    sources:
        - filesystem
        - coverart: release releasegroup
        - itunes
        - '*'
    auto: yes
    art_filename: cover.jpg
    store_source: yes

bucket:
    bucket_year: ['1930s', '1940s', '1950s', '1960s', '1970s', '1980s', '1990s', '2000s', '2010s', '2020s']

paths:
    default: %bucket{$year}/$year - $albumartist - $album/$track - $title
    comp: %bucket{$year}/$year - $albumartist - $album/$track - $artist - $title

replaygain:
    backend: gstreamer
    auto: no
    overwrite: no
    targetlevel: 89

follow:
    auto: no
    email: me@something.ext
    password: XXX
    userid: XXX

wlg:
    auto: no
    force: no
    count: 3
    separator: ', '

web:
    host: 0.0.0.0
    cors: '*'
    
discogs:
    user_token: XXX
```

### PR DESCRIPTION
**Pull request by sampsyo**
This is an attempt to fix #3773 by more aggressively normalizing values for dbcore's `String` type.

### CODE REVIEW

**Comment by wisp3rwind** [comment]
That looks a lot like

https://github.com/beetbox/beets/blob/52ca0cbfdecc1bc22139234acf0b666a13d4a5de/beets/dbcore/types.py#L89-L91

should be adressed, right? Does anyone (in particular @geigerzaehler according to `git blame`) remember why that comment was added in the first place? Maybe, given that this PR has no CI failures, there's no issue with coercing all values to their `model_type` anymore?

In fact, maybe `String.normalize()` should ultimately be more strict, and assert that its values are actually `six.text_type`, rather than just converting them. Pretty much any value can successfully be converted to a string, so there's no real type checking otherwise.

**Comment by sampsyo** [comment]
Absolutely, yeah, that would be the thing to fix. Here's the context as best as I can remember. Before we added the type system, lots of parts of the codebase were very loosey-goosey about how they used various fields. This was a growing problem, of course, which is why we added the types in the first place. But we were nervous about all that existing code being that was so loosey-goosey about the types. The idea was to proceed in two phases:

- Get the type system into place, but don't disrupt existing code—behave the same way as the type-free old system.
- Flip the switch to turn on automatic conversions.

You could imagine adding a third phase to this, as @wisp3rwind suggested: instead of silently converting things like `item.title = 123`, start throwing errors. But it seems like we stalled out after the first phase. Getting more aggressive about this would be the right thing to do, however.

File: beets/dbcore/types.py [hunk 1]
```diff
@@ -207,6 +207,12 @@ class String(Type):
     sql = u'TEXT'
     query = query.SubstringQuery
 
+    def normalize(self, value):
+        if value is None:
+            return self.model_type()
```
Comment (wisp3rwind) [inline beets/dbcore/types.py:212]: I think it would be cleaner to return `self.null` here (even though that is just `self.model_type()` for `String`.

**Review by wisp3rwind — COMMENTED** [review]
I just tried to run the tests with the change mentioned in the `TODO` note above. There's one test failure, namely `test_format_flex_field_bytes` in `test_dbcore`. The issue is that `types.Default` then forces all flexattrs to unicode, but the test expects to be able to write utf8 bytes to the database. I'm pretty sure that forcing unicode here would break flexattr usage by some plugins. And since the database actually stores bytes rather than unicode objects (right?), it seems sensible to be able to write raw bytes.

So it might indeed by best to only coerce `String`s to their `model_type`, not all fields.

File: beets/dbcore/types.py (hunk 1 above)
Comment (sampsyo) [reply beets/dbcore/types.py:212]: Yeah, you're absolutely right about that.

**Comment by sampsyo** [comment]
Intriguing! Thanks for trying that out. Let's keep the dream alive—maybe we can work around that limitation by, for example, allowing fields without a declared type to remain dynamically typed and not coerced. But for now, we can make this more conservative change for strings only.

**Pull request merged**